File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode flatten_01(TreeNode node) {
        if(node == null || (node.left == null && node.right == null)) return node;
    
        TreeNode leftTail = flatten_01(node.left);
        TreeNode rightTail = flatten_01(node.right);
        
        if(leftTail != null) {
            leftTail.right = node.right;
            node.right = node.left;
            node.left = null;
        }
        
        return rightTail != null ? rightTail : leftTail;
    }
    
    public void flatten(TreeNode root) {
        flatten_01(root);
    }
}

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* flatten_01(TreeNode* node) {
        //Base Case to handle Leafs
        if(!node || (!node->left && !node->right)) return node;
        //Left and Right recursive Call;
        TreeNode* leftTail = flatten_01(node->left);
        TreeNode* rightTail = flatten_01(node->right);
        
        //Do the attachments
        if(leftTail) {
            leftTail->right = node->right;
            node->right = node->left;
            node->left = nullptr;
        }
        
        return rightTail != nullptr ? rightTail : leftTail;
    }
    
    void flatten(TreeNode* root) {
        flatten_01(root);
    }
};